src/System.Collections/src/System/Collections/BitArray.cs

```diff
@@ -366,23 +366,32 @@ public int Length
 
                 if (value > m_length)
                 {
-                    // clear high bit values in the last int
-                    int last = GetArrayLength(m_length, BitsPerInt32) - 1;
-                    int bits = m_length % 32;
-                    if (bits > 0)
-                    {
-                        m_array[last] &= (1 << bits) - 1;
-                    }
-
-                    // clear remaining int values
-                    Array.Clear(m_array, last + 1, newints - last - 1);
+                    TrimUpperBits();
                 }
 
                 m_length = value;
-                _version++;
             }
         }
 
+        private void TrimUpperBits()
+        {
+            // clear high bit values in the last int
+            int last = GetArrayLength(m_length, BitsPerInt32) - 1;
+            int bits = m_length % 32;
+            if (bits > 0)
+            {
+                m_array[last] &= (1 << bits) - 1;
+            }
+
+            // clear remaining int values
+            if (last + 1 < m_array.Length)
+            {
+                Array.Clear(m_array, last + 1, m_array.Length - last - 1);
+            }
+
+            _version++;
```

Comment:
@mareklinka - there's a proposal out there to add shift operators.
@MichalStrehovsky - we need to sanitize after setting the length, since otherwise it is visible.